### PR TITLE
ISender interface, derived class BitSender

### PR DESCRIPTION
**Pull request by Riyria-was-taken**
- implemented interface for data sending - ISender.h as well as one class fulfiling it - BitSender.*
- moved part of responsibilities (encoding) from IAudioConverter.h, BitAudioConverter.h and moved data sending implementation from Echo.c to ISender.h, BitSender.*
- run clang-format on AudioInput.h, AudioStream.h, Echo.cc and tmp_test.cpp

### CODE REVIEW

**Comment by Riyria-was-taken** [comment]
Note: tests in another PR

File: libecho/CMakeLists.txt [hunk 1]
```diff
@@ -15,20 +15,21 @@ endif ()
 set(SRC_DIR "src" CACHE PATH "Source files path")
 set(API_DIR "include" CACHE PATH "Public API files path")
 set(SRC_FILES
-        ${SRC_DIR}/Echo.cc
+	${SRC_DIR}/Echo.cc
+	${SRC_DIR}/AudioFormatFactory.h
+	${SRC_DIR}/AudioFormatFactory.cc
```
Comment (Lorak-mmk) [inline libecho/CMakeLists.txt]: Please fix indentation

File: libecho/src/BitSender.cc [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+#include <cmath>
+
+const QAudioFormat::Endian OUTPUT_BYTEORDER = QAudioFormat::LittleEndian;
+const int OUTPUT_CHANNEL_COUNT = 1;
+const char *OUTPUT_CODEC = "audio/pcm";
+const int OUTPUT_SAMPLERATE = 44100;
+const int OUTPUT_SAMPLESIZE = sizeof(SampleType);
+const QAudioFormat::SampleType OUTPUT_SAMPLETYPE = QAudioFormat::SignedInt;
+/**< QAudioFormat settings we use while sending audio. */
```
Comment (Lorak-mmk) [inline libecho/src/BitSender.cc]: did you check if this comment applies only to OUTPUT_SAMPLETYPE or to all the consts above?

File: libecho/src/BitSender.cc [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+static constexpr size_t SOUNDS_PER_BYTE =
+    8; /**< States the number of sounds we need to play to encode a byte of data. */
+
+void BitSender::send(const std::vector<uint8_t> &buffer) {
+    auto encoded = encode(buffer);
+    output->startStream();
```
Comment (Lorak-mmk) [inline libecho/src/BitSender.cc]: We discussed moving startStream/stopStream to constructor/destructor. What do you think about it?

File: libecho/tests/main.cpp [hunk 4]
```diff
@@ -1,9 +1,7 @@
-#include "Echo.h"
-
 #include <gtest/gtest.h>
 
 int main(int argc, char **argv) {
     ::testing::InitGoogleTest(&argc, argv);
-    Echo::initEcho(argc, argv);
+    //     Echo::initEcho(argc, argv);
```
Comment (Lorak-mmk) [inline libecho/tests/main.cpp]: I don't think anything will work if this is not called.

**Review by Lorak-mmk — CHANGES_REQUESTED** [review]

File: libecho/src/BitSender.cc (hunk 3 above)
Comment (Riyria-was-taken) [reply libecho/src/BitSender.cc]: kki

File: libecho/src/BitSender.cc (hunk 2 above)
Comment (Riyria-was-taken) [reply libecho/src/BitSender.cc]: It doesn't appear in doxygen documentation as this is .cc file.

**Review by PawelA — APPROVED** [review]

File: libecho/src/BitSender.cc (hunk 2 above)
Comment (Lorak-mmk) [reply libecho/src/BitSender.cc]: Then why use Doxygen syntax?

**Review by Lorak-mmk — APPROVED** [review]

File: libecho/src/BitSender.cc [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+const QAudioFormat::Endian OUTPUT_BYTEORDER = QAudioFormat::LittleEndian;
+const int OUTPUT_CHANNEL_COUNT = 1;
+const char *OUTPUT_CODEC = "audio/pcm";
+const int OUTPUT_SAMPLERATE = 44100;
+const int OUTPUT_SAMPLESIZE = 8;
+const QAudioFormat::SampleType OUTPUT_SAMPLETYPE = QAudioFormat::UnSignedInt;
+/**^ QAudioFormat settings we use while sending audio. */
+
+static constexpr size_t SOUNDS_PER_BYTE =
+    8; /**< States the number of sounds we need to play to encode a byte of data. */
```
Comment (Lorak-mmk) [inline libecho/src/BitSender.cc]: 8 in next line looks bad, put comment before this line, and 8 with the rest of declaration

File: libecho/tests/tmp_test.cpp [hunk 6]
```diff
@@ -1,9 +1,52 @@
...
+    std::vector<uint8_t> vbuffer;
+    for (int i = 0; i < 256; i++) {
+        vbuffer.push_back(i);
+    }
+    BitSender s(200, 14000, 15000);
+    s.send(vbuffer);
+    sleep(10);
+}
+    
+
```
Comment (Lorak-mmk) [inline libecho/tests/tmp_test.cpp]: I'd suggest deleting this whole file

**Review by Lorak-mmk — APPROVED** [review]

File: libecho/src/BitSender.cc (hunk 2 above)
Comment (Riyria-was-taken) [reply libecho/src/BitSender.cc]: 不知道

**Pull request merged**